### PR TITLE
Make write an operator as part of the execution plan

### PR DESCRIPTION
**Pull request by jianoaix**
## Why are these changes needed?
Today the "write" is not captured as an operator in the execution plan: it just fully executes the plan and then launches Ray tasks to perform the writing.

By make it a write operator, we will have a cleaner/unified model to handle how "write" is supported, enable query optimization for write (e.g. read->map->write can be fused into one operator node) for better performance.

Note: supporting this write operator in the new query plan will be in a follow-up PR.

Benchmarked on a single node:
```
import ray
import time

def make_ds(size_gb: int, parallelism: int):
    # Dataset of 10KiB tensor records.
    total_size = 1024 * 1024 * 1024 * size_gb
    record_dim = 1280
    record_size = record_dim * 8
    num_records = int(total_size / record_size)
    dataset = ray.data.range_tensor(num_records, shape=(record_dim,), parallelism=parallelism)
    print("Created dataset", dataset, "of size", dataset.size_bytes())
    return dataset

sizes = [1, 5, 10]
parallelism = [1, 5, 10, 20, 100]

perf = {}

for sz in sizes:
    for p in parallelism:
        path = f"/tmp/write/test-{sz}-{p}"
        time_start = time.perf_counter()
        ds = make_ds(sz, p).map_batches(lambda x: x).write_numpy(path)
        perf[path] = time.perf_counter() - time_start

print(perf)
```

Before:
```
{'/tmp/write/test-1-1': 9.953020322000157, '/tmp/write/test-1-5': 3.428301415000533, '/tmp/write/test-1-10': 4.723238538999794, '/tmp/write/test-1-20': 5.163609835999523, '/tmp/write/test-1-100': 4.33488237399979, '/tmp/write/test-5-1': 47.80200573299953, '/tmp/write/test-5-5': 57.3792794459996, '/tmp/write/test-5-10': 21.089490562999345, '/tmp/write/test-5-20': 49.24927898700025, '/tmp/write/test-5-100': 47.518018443000074, '/tmp/write/test-10-1': 182.66500996800005, '/tmp/write/test-10-5': 139.72414283000035, '/tmp/write/test-10-10': 129.0919650550004, '/tmp/write/test-10-20': 109.1896361629997, '/tmp/write/test-10-100': 61.62153517600018}

real    14m35.661s
user    0m13.973s
sys     0m2.241s
```

After:
```
{'/tmp/write/test-1-1': 14.615472456999669, '/tmp/write/test-1-5': 3.4766328169998815, '/tmp/write/test-1-10': 2.675365926000268, '/tmp/write/test-1-20': 2.3434579160002613, '/tmp/write/test-1-100': 2.7332164729996293, '/tmp/write/test-5-1': 29.679511864000233, '/tmp/write/test-5-5': 18.722395105999567, '/tmp/write/test-5-10': 18.884924998000315, '/tmp/write/test-5-20': 20.06004047999977, '/tmp/write/test-5-100': 18.469305754000743, '/tmp/write/test-10-1': 160.16473960199983, '/tmp/write/test-10-5': 47.103051861000495, '/tmp/write/test-10-10': 39.20541486499951, '/tmp/write/test-10-20': 38.90085198700035, '/tmp/write/test-10-100': 40.142852582999694}

real    7m39.632s
user    0m9.006s
sys     0m2.634s
```

The write fusion generally makes it faster. And it took only half of the time with fusion to complete this entire benchmark.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/execution/legacy_compat.py
Comment (ericl) [inline python/ray/data/_internal/execution/legacy_compat.py:150]: Could we define a TaskContext class and put the index there? Otherwise, we may have to keep adding fields over time and it will get kind of messy.

File: python/ray/data/dataset.py
Comment (ericl) [inline python/ray/data/dataset.py:2685]: This isn't the same semantics right? on_write_complete/failed should be called on the driver only, not once per block.

File: python/ray/data/datasource/datasource.py
Comment (ericl) [inline python/ray/data/datasource/datasource.py:53]: Is it possible to make this change backwards compatible, like we do for the reader API?

**Review by ericl — COMMENTED** [review]
This is a pretty big PR, shall we split it up?

File: python/ray/data/datasource/datasource.py
Comment (jianoaix) [reply python/ray/data/datasource/datasource.py]: Added a new method for direct writing out, and deprecated `do_write()`.

File: python/ray/data/dataset.py
Comment (ericl) [reply python/ray/data/dataset.py:2635]: You'd need these to implement commit semantics / rollback changes on failure, for example. They are part of the DataSource public API, so we shouldn't change them without a separate API discussion.

Comment (c21) [reply python/ray/data/dataset.py:2635]: I feel it's too rush for 2.3 release if we need to change `Datasource` API at this moment, and agree that separate discussion needs to happen if we change `Datasource` API here. How about we taking a non-intrusive approach to not change `Datasource` API at all?

How about we are doing this:

```py
if hasattr(datasource, "_direct_write"):
  def transform(blocks: Iterable[Block], ctx) -> Iterator[Block]:
    yield from datasource._direct_write(blocks, ctx, **write_args)
  self._plan = self._plan.with_stage(transform, ...)
  write_results = self._plan.execute().get_blocks()
else:
  # Old existing code path, or we implement a AllToAllStage, that should give us Block and BlockMetadata
  write_results = ...

try:
  progress.block_until_complete(write_results)
  datasource.on_write_complete(ray.get(write_results))
except Exception as e:
  datasource.on_write_failed(write_results, e)
```

Comment (jianoaix) [reply python/ray/data/dataset.py:2635]: I think there is no issue blocked by not having write operator/fusion at the moment, so I'd fix it properly instead of having a half baked feature (2.3 is not the ultimate cut to catch anyway).

Comment (ericl) [reply python/ray/data/dataset.py:2635]: Yeah, I think we don't need this for 2.3.

Properly supporting the original callback semantics should be possible within the PhysicalOperator framework I believe, though you might need a custom subclass of the map operator.

File: python/ray/data/datasource/file_based_datasource.py
Comment (ericl) [inline python/ray/data/datasource/file_based_datasource.py:328]: Lets have do_write() delegate to direct_write() so that the code paths are unified?

File: python/ray/data/datasource/file_based_datasource.py
Comment (ericl) [inline python/ray/data/datasource/file_based_datasource.py:262]: Why not following the same API pattern as the Reader (have a WriteTask class that is created)?

File: python/ray/data/dataset.py
Comment (jianoaix) [reply python/ray/data/dataset.py:2635]: I'm thinking we may use the results of the write tasks to indicate if the entire write operation is complete or failed this way:
- The write tasks can have an output anyway: i.e. the `transform_fn` will be mapping `Iterator[Blocks]` to `WriteResult`. If the task succeeds, `WriteResult` is "ok", otherwise it's the error object from the task.
- The `write_xx()` still returns None, but internally it will run the plan with the write as the last operator; the results of executing that plan is a Dataset comprised of `List[WriteResult]`.
- We don't expose this dataset `List[WriteResult]` from `write_xx()`, but internally use it as an indicator to run these two callbacks: if all elements are "ok", `on_write_complete`, otherwise `on_write_failed`.

This seems address the callback semantics?

Comment (ericl) [reply python/ray/data/dataset.py:2635]: That makes sense, assuming the result isn't literally "ok" but the user-defined write return value.

Comment (jianoaix) [reply python/ray/data/dataset.py]: Hmm, it looks we cannot have the `transform_fn` return the error -- it has to raise the error, in order to trigger the failed tasks retry (we caught this in unit test: https://buildkite.com/ray-project/oss-ci-build-pr/builds/11123).

The right place seems to be the PhysicalOperator as you suggested, as they launch and handle tasks directly. We may add `on_operator_failed()` and `on_operator_complete()` to its interface, and plumb the callback functions to the operator similar to how we plumb the `transform_fn`.

Comment (ericl) [reply python/ray/data/dataset.py]: It's probably ok to raise the error still I think with the proposed strategy, is there a reason why you had to return them?

Comment (jianoaix) [reply python/ray/data/dataset.py]: Returning (instead of raising) errors was to make the above approach (the write operator producing a "status Dataset"), but this is not working for retry.

Now as we agree to add interface to PhysicalOperator to handle this, then we will still raise the error, so Ray will know it should retry. I'm going to change it to this way.

Comment (ericl) [reply python/ray/data/dataset.py]: I think the status Dataset approach still works right? Errors can get raised as usual, so retires will work. In the write function here we can add a try catch around the entire Dataset execution operation, and call the right data source callbacks if the entire thing fails / succeeds.

Comment (ericl) [reply python/ray/data/dataset.py]: Oh, do you mean it's hard to provide List[ObjectRef[WriteResult]] to the error callback? I think we can probably deprecate this part of the API for the new backend (pass empty list). Afaik it's not really useful and would be best effort in any scenario for a distributed system.

Comment (jianoaix) [reply python/ray/data/dataset.py]: Right, we will not know the resulting status of each individual tasks, which I think may be used for error handling (rollback etc.) in the callback function. If we just try catch the entire execution, and just produce an empty list (as the result of executing write operator), then yes it should be able to work.

Comment (ericl) [reply python/ray/data/dataset.py]: How about we go for that then? It seems the simplest approach, and we can also fix the new API to not have the list of error refs.

Comment (jianoaix) [reply python/ray/data/dataset.py]: Sounds good, I'll update the PR.

File: python/ray/data/datasource/file_based_datasource.py
Comment (jianoaix) [reply python/ray/data/datasource/file_based_datasource.py:331]: Sounds good, done.

File: python/ray/data/datasource/file_based_datasource.py
Comment (jianoaix) [reply python/ray/data/datasource/file_based_datasource.py:262]: For write, I think that's adding multiple layers of unnecessary indirections. We added Reader was due to the need of memory estimation, but we do not need thing like that for the write here.

**Comment by jianoaix** [comment]
Benchmarked on a single node:
```
import ray
import time

def make_ds(size_gb: int, parallelism: int):
    # Dataset of 10KiB tensor records.
    total_size = 1024 * 1024 * 1024 * size_gb
    record_dim = 1280
    record_size = record_dim * 8
    num_records = int(total_size / record_size)
    dataset = ray.data.range_tensor(num_records, shape=(record_dim,), parallelism=parallelism)
    print("Created dataset", dataset, "of size", dataset.size_bytes())
    return dataset

sizes = [1, 5, 10]
parallelism = [1, 5, 10, 20, 100]

perf = {}

for sz in sizes:
    for p in parallelism:
        path = f"/tmp/write/test-{sz}-{p}"
        time_start = time.perf_counter()
        ds = make_ds(sz, p).map_batches(lambda x: x).write_numpy(path)
        perf[path] = time.perf_counter() - time_start

print(perf)
```

Before:
```
{'/tmp/write/test-1-1': 9.953020322000157, '/tmp/write/test-1-5': 3.428301415000533, '/tmp/write/test-1-10': 4.723238538999794, '/tmp/write/test-1-20': 5.163609835999523, '/tmp/write/test-1-100': 4.33488237399979, '/tmp/write/test-5-1': 47.80200573299953, '/tmp/write/test-5-5': 57.3792794459996, '/tmp/write/test-5-10': 21.089490562999345, '/tmp/write/test-5-20': 49.24927898700025, '/tmp/write/test-5-100': 47.518018443000074, '/tmp/write/test-10-1': 182.66500996800005, '/tmp/write/test-10-5': 139.72414283000035, '/tmp/write/test-10-10': 129.0919650550004, '/tmp/write/test-10-20': 109.1896361629997, '/tmp/write/test-10-100': 61.62153517600018}

real    14m35.661s
user    0m13.973s
sys     0m2.241s
```

After:
```
{'/tmp/write/test-1-1': 14.615472456999669, '/tmp/write/test-1-5': 3.4766328169998815, '/tmp/write/test-1-10': 2.675365926000268, '/tmp/write/test-1-20': 2.3434579160002613, '/tmp/write/test-1-100': 2.7332164729996293, '/tmp/write/test-5-1': 29.679511864000233, '/tmp/write/test-5-5': 18.722395105999567, '/tmp/write/test-5-10': 18.884924998000315, '/tmp/write/test-5-20': 20.06004047999977, '/tmp/write/test-5-100': 18.469305754000743, '/tmp/write/test-10-1': 160.16473960199983, '/tmp/write/test-10-5': 47.103051861000495, '/tmp/write/test-10-10': 39.20541486499951, '/tmp/write/test-10-20': 38.90085198700035, '/tmp/write/test-10-100': 40.142852582999694}

real    7m39.632s
user    0m9.006s
sys     0m2.634s
```

The write fusion generally makes it faster. And it took only half of the time with fusion to complete this entire benchmark.

File: python/ray/data/dataset.py
Comment (ericl) [inline python/ray/data/dataset.py:2654]:
```suggestion
        def transform(blocks: Iterable[Block], ctx, fn) -> Iterable[Block]:
```

File: python/ray/data/datasource/datasource.py
Comment (ericl) [inline python/ray/data/datasource/datasource.py:54]:
```suggestion
    def write(
```

File: python/ray/data/datasource/datasource.py
Comment (ericl) [inline python/ray/data/datasource/datasource.py:365]: I don't think we need to change this API.

File: python/ray/data/dataset.py
Comment (ericl) [inline python/ray/data/dataset.py:2668]: Let's call ray.get() here.

File: python/ray/data/datasource/file_based_datasource.py
Comment (ericl) [inline python/ray/data/datasource/file_based_datasource.py:347]: Nothing is calling do_write() any more right? Should we just remove it for our internal datasources?

File: python/ray/data/datasource/mongo_datasource.py
Comment (ericl) [inline python/ray/data/datasource/mongo_datasource.py:76]: Not used.

File: python/ray/data/tests/test_dataset.py
Comment (ericl) [inline python/ray/data/tests/test_dataset.py:5524]: This isn't testing that we support do_write() through the Dataset API. For proper backwards compatibility, we need to (1) check if the datasource doesn't have direct write implemented, and (2) in that case, Dataset needs to fallback to using do_write() as an AllToAll stage instead of a OneToOne.

**Review by ericl — CHANGES_REQUESTED** [review]
Main comment is that backwards compatibility needs to be preserved for legacy datasources that implement only do_write(); those should work when you `write_datasource()`.

File: python/ray/data/datasource/file_based_datasource.py
Comment (jianoaix) [reply python/ray/data/datasource/file_based_datasource.py]: done.

File: python/ray/data/tests/test_dataset.py
Comment (jianoaix) [reply python/ray/data/tests/test_dataset.py]: I'm checking the presence of "write" attribute to decide whether this should fallback to the legacy path.

File: python/ray/data/dataset.py
Comment (ericl) [inline python/ray/data/dataset.py:2686]: Can we log a deprecation warning here (add Todo to hard deprecate later?)

**Review by ericl — APPROVED** [review]

File: python/ray/data/dataset.py
Comment (jianoaix) [reply python/ray/data/dataset.py:2686]: Done.

Comment (ericl) [reply python/ray/data/dataset.py:2686]: I meant for the entire legacy do_write API.

Comment (jianoaix) [reply python/ray/data/dataset.py:2686]: Ah, done. I also added deprecation message to the API already.

Should RAY_DATASET_FORCE_LOCAL_METADATA be deprecated as well? There is a TODO but unclear what's the conditions to actually remove it.

Comment (ericl) [reply python/ray/data/dataset.py:2686]: I think it's safe to remove this now, it was only an escape hatch I believe and not documented.

**Review by ericl — APPROVED** [review]

File: python/ray/data/dataset.py
Comment (jianoaix) [reply python/ray/data/dataset.py:2686]: I'll leave it to a followup cleanup.

**Pull request merged**
